src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -120,39 +122,28 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig, protected
     }
   }
 
-  private def getFirewallRule(googleProject: String, firewallRuleName: String): Future[Unit] = {
-    Future {
-      val request = compute.firewalls().get(googleProject, firewallRuleName)
-      executeGoogleRequest(request) // returns a Firewall
-      ()
-    }.recoverWith {
-      case e: GoogleJsonResponseException =>
-        if (e.getStatusCode == 404)
-          throw FirewallRuleNotFoundException(googleProject, firewallRuleName)
-        throw CallToGoogleApiFailedException(googleProject, firewallRuleName, e.getStatusCode, e.getDetails.getMessage)
-    }
+  private def getFirewallRule(googleProject: GoogleProject, firewallRuleName: String): Future[Unit] = {
+    val request = compute.firewalls().get(googleProject, firewallRuleName)
+    executeGoogleRequestAsync(googleProject, firewallRuleName, request).recover {
+      case CallToGoogleApiFailedException(_, _, 404, _) =>
+        // throw FirewallRuleNotFoundException in case of 404 errors
+        throw FirewallRuleNotFoundException(googleProject, firewallRuleName)
+    }.void
   }
 
-
   /* Adds a firewall rule in the given google project. This firewall rule allows ingress traffic through a specified port for all
      VMs with the network tag "leonardo". This rule should only be added once per project.
     To think about: do we want to remove this rule if a google project no longer has any clusters? */
   private def addFirewallRule(googleProject: GoogleProject): Future[Unit] = {
-      // Create an Allowed object that specifies the port and protocol of rule
-      val allowed = new Allowed().setIPProtocol(proxyConfig.jupyterProtocol).setPorts(List(proxyConfig.jupyterPort.toString).asJava)
-      // Create a firewall rule that takes an Allowed object, a name, and the network tag a cluster must have to be accessed through the rule
-      val googleFirewallRule = new Firewall()
-        .setName(dataprocConfig.clusterFirewallRuleName)
-        .setTargetTags(List(dataprocConfig.clusterNetworkTag).asJava)
-        .setAllowed(List(allowed).asJava)
-      val request = compute.firewalls().insert(googleProject, googleFirewallRule)
-    Future {
-      executeGoogleRequest(request) // returns a ComputeOperation
-      ()
-    }.recoverWith {
-      case e: GoogleJsonResponseException =>
-        throw CallToGoogleApiFailedException(googleProject, dataprocConfig.clusterFirewallRuleName, e.getStatusCode, e.getDetails.getMessage)
-    }.mapTo[Unit]
+    // Create an Allowed object that specifies the port and protocol of rule
+    val allowed = new Allowed().setIPProtocol(proxyConfig.jupyterProtocol).setPorts(List(proxyConfig.jupyterPort.toString).asJava)
+    // Create a firewall rule that takes an Allowed object, a name, and the network tag a cluster must have to be accessed through the rule
+    val googleFirewallRule = new Firewall()
```

Comment:
This is constant so we should move it outside the function

Reply:
👍